### DISCUSSION TITLE
This project is awesome, but over my head - how can I get started?

### DISCUSSION THREAD

**Discussion by MoSattler** [epicweb-dev/epic-stack#130] [General]
I love that Epic Stack is tackling the analysis paralysis problem, and it is packed with many advanced features like [this](https://github.com/epicweb-dev/epic-stack/commit/560585d0ced28165e62c6ac4df89bc23e4b5d4ba) , which is super nice, but can be intimidating at times. I often don't fully understand the code or its purpose, and I don't dare to touch it or make changes, cause I don't fully understand the implications. 

This might be more of a personal challenge, but I'd still was wondering if anyone has any advice on how to take advantage of all the goodness this starter-pack brings, while being able to confidently develop my product with it? Should I just get started and not worry about it? Are things usually safe to tweak or remove if I don't need them?

**Comment by kentcdodds** [comment]
Great thoughts. My hope is to address this with superb documentation. It's going to take time. Definitely don't feel bad that some things are over your head. That's my fault, not yours. Feedback on complicated subjects is welcome and will help me prioritize my efforts.

**Comment by kentcdodds** [comment]
Oh, and definitely feel free to tweak or remove anything you like. That's the benefit of using a generator like this

**Comment by juanojeda** [comment]
Hey @kentcdodds, firstly and most importantly, thanks for taking the time to put this project together. I can really see that it will speed up a lot of developers, and make it much easier to GSD.  I can also see that a lot of work has gone into this, so once again, thanks.

I would, however, like to second the sentiment that it's hard to know where to start. I'm currently assessing this project against a couple of other web project starters, but it's hard to know what parts of the code are part of Epic Stack, and which are just examples of what you _can_  do.

For example - Presumably, [email.server.ts](https://github.com/epicweb-dev/epic-stack/blob/main/app/utils/email.server.ts) is a part of Epic Stack, but on [line 34](https://github.com/epicweb-dev/epic-stack/blob/main/app/utils/email.server.ts#L34C26-L34C26) there's a `const` defined that is clearly just for the example. There are probably other examples like this.

Just wondering if I'm missing a step, or if this is something that is likely to be improved in the future?

Edit:
I've just found [this topic](https://github.com/epicweb-dev/epic-stack/discussions/170) that seems to be related to what I mean too. Apologies if this comment is in the wrong place.

**Reply by sergiodxa** [reply]
The const should probably come from a `config.server.ts` file where it can centralize those configurations.

**Reply by kentcdodds** [reply]
That's a fair point. I think I'd be willing to look at doing something like that.

**Comment by vinstah** [comment]
Hey epic-stack will help you get started with the config but when it comes to creating something with it think of smallest possible win, like changing something in root.tsx or global search epic and see what comes up

**Comment by DigitalGoldfish** [comment]
My personal experience: I started a project for a client with the epic stack a month ago having no previous experience with remix but a working knowledge of react. Now a month later there are still parts of the epic stack/remix I have barely looked at much less fully understand. Some others topics I figured out when I needed them (e.g. routing, file uploads, form handling, ...) and it wasn't that hard once I focused at them one by one while building my project without trying to understand everything at the same time. 

So my recommendation would be to just get started building whatever you want to build going step by step and don't try to understand everything at once.

**Reply by kentcdodds** [reply]
That's great to hear! Thank you for sharing your experience!

**Reply by MoSattler** [reply]
@DigitalGoldfish Thanks for the advice. My concerns about using code I don't understand are: a) I can't be certain if anything might unintentionally appear to users; b) if a bug arises, it may be harder to find and I'd have to work with unfamiliar code to fix it; c) making even slight changes could unexpectedly break something elsewhere, and I wouldn't know how or why. Additionally, as @juanojeda noted, it's not always easy to tell whether specific code is part of the underlying stack or an example, which adds to the complexity.

**Reply by MoSattler** [reply]
To clarify, I haven't actually encountered these issues in practice. They are simply concerns that prevent me from confidently investing time and money into serious projects with this, the way you did, @DigitalGoldfish. So maybe those concerns are unfounded.

**Reply by kentcdodds** [reply]
I understand the concern.

Keep in mind that some of the things you mentioned can happen when you use libraries as well. The difference in that case is you're at the mercy of the library authors to fix issues, whereas with the EpicStack you can at least invest time in understanding the code and make a change yourself.

There are trade-offs there, and I do plan on packaging up more of the Epic Stack as packages (like I did with the totp functionality #361).

**Comment by MoSattler** [comment]
> Keep in mind that some of the things you mentioned can happen when you use libraries as well. The difference in that case is you're at the mercy of the library authors to fix issues, whereas with the EpicStack you can at least invest time in understanding the code and make a change yourself.

I completely understand where you're coming from, I just want to share another view point on this:
When I'm embarking on a new project, my primary focus is often on the business logic and shipping the vision as swiftly as possible. So I lean towards the convenience and shared responsibility that libraries offer, rather than managing all the code myself (fixing bugs, spotting vulnerabilities, updating, etc.). This approach allows me to concentrate on the core business logic.

To provide an example of another related of my concerns: I'm currently using the blues-stack for several projects. Every few weeks, I find myself checking its repo for any updates and trying to incorporate those changes into my projects that use that stack. While the blues-stack is less opinionated, this can still be a challenge at times. I suspect that if I had to do the same with the epic-stack, which is more opinionated and complex for valid reasons, the process might become even more intricate.

**Reply by kentcdodds** [reply]
Yep, I can definitely appreciate that. It's a trade-off. Not sure what actionable things can come out of this GitHub discussion that I'm not already actively working on though 🤷‍♂️

**Reply by sergiodxa** [reply]
> Every few weeks, I find myself checking its repo for any updates and trying to incorporate those changes into my projects that use that stack

This is the trade off of using a starter vs a framework.

On the starter you own the code, but it also means that getting updates to that code that now you own is a manual work, specially if you did changes.

While on a framework, you just update a dependency to get new versions, but you also don't own the code so you can't as easily do changes.

I think before choosing EpicStack people should choose which tradeoff they want to use

1. Own the code
2. Easy to update

Some parts of EpicStack could also be packages (e.g. the totp package) but not everything makes sense to be a package, and whatever is put in a package will be code the app developer will not own anymore.

**Reply by kentcdodds** [reply]
Yep, and when you make this decision, consider your alternatives. I don't think there's anything currently that can do quite as much for you as the Epic Stack can in many areas (though if you're willing to use a non-js framework you have options, but there are other trade-offs there).

So your choice is:

1. A starter that gets you going with a good start
2. Building it all yourself and owning that code anyway

🤷‍♂️

**Reply by vinstah** [reply]
While it would be cool to see epic-stack turn into something like nuxt where you can choose what to use, I help the upgrading path by keeping my application in a separate folder, that way I can create a new instance of epic-stack and copy the folder over then update the git remote to push a new release

**Reply by MoSattler** [reply]
Thanks, everyone, for your insights and highlighting these these subtleties. I recognize it's a balancing act and a trade-off. Perhaps I'm actually more in need of a framework than a starter like epic-stack. Still, I'll definitely keep following things and maybe play around with it a bit more.

**Reply by kentcdodds** [reply]
I honestly think one of the primary uses people will have out of the Epic Stack is not in starting new projects with it but in referencing it as needed when you build out your own app. So it's fine if you don't start with the project. You can still learn a lot just by looking at how I do things in here :)

**Comment by MoSattler** [comment]
Hey everyone, just wanted to drop in and give a small update on my journey with the epic-stack. I decided to take the plunge and move a blues-stack project over to epic-stack. Admittedly, there was a learning curve at the start, but once I began to get the hang of it, things started falling into place.

One of my initial worries was about moving from Postgres to SQLite, but much to my surprise, it went easier than I thought. A big shoutout to [this article](https://kentcdodds.com/blog/i-migrated-from-a-postgres-cluster-to-distributed-sqlite-with-litefs) by @kentcdodds which helped me a lot through the process.

Right now, I'm not using all the features of the epic-stack, but the parts I've taken on, and the patterns I've found have really made my own code better, the performance of my applications, and even other things like CI speed.

In short, I'm really happy! I'm officially fan and am excited to keep working with it. To everyone who gave advice and thoughts earlier, a big thanks! Cheers! 🙌

**Reply by kentcdodds** [reply]
I'm glad to hear it! Thank you for your contributions as well ☺️

**Comment by fedorinoGore** [comment]
It is 2025. 
And the question is still here. 
Where I can see the tutorials of real-life applications build with EPIC stack? 
There tens of NEXT.js tutorials that showing step by step how to create original web-app or SaaS from scratch. 
An only one for EPIC - Kent’s course, Epic Web Dev...

**Reply by MoSattler** [reply]
In the end I used it to built a few apps for production. Worked great for me.
